src/opnsense/scripts/filter/lib/alias/__init__.py

```diff
@@ -164,36 +164,51 @@ def resolve(self, force=False):
             :return: string
         """
         if not self._resolve_content:
+
             if self.expired() or self.changed() or force:
-                if os.path.isfile(self._filename_alias_content):
+                alias_file_exists = os.path.isfile(self._filename_alias_content)
+                if alias_file_exists:
                     try:
                         undo_content = open(self._filename_alias_content, 'r').read()
                     except UnicodeDecodeError:
                         undo_content = ""
                 else:
                     undo_content = ""
+                address_parser = self.get_parser()
+                if os.path.isfile(self._filename_alias_hash):
+                    try:
+                        hash_contents = open(self._filename_alias_hash, 'r').read()
+                    except UnicodeDecodeError:
+                        hash_contents = ""
+                else:
+                    hash_contents = ""
+                unique_id = self.uniqueid()
+                syslog.syslog(syslog.LOG_ERR, 'add %s (%s, %s, %s/%s)' % (self._name, self.expired(), self.changed(), force, hash_contents, unique_id))
                 try:
                     self._resolve_content = self.pre_process()
-                    address_parser = self.get_parser()
                     if address_parser:
                         for item in self.items():
                             for address in address_parser.iter_addresses(item):
                                 self._resolve_content.add(address)
                         # resolve hostnames (async) if there are any in the collected set
                         self._resolve_content = self._resolve_content.union(address_parser.resolve_dns())
-                    # Always save last recorded content to disk, also when we're not responsible for the alias
-                    # so we can use cached results when reloading a single alias.
-                    with open(self._filename_alias_content, 'w') as f_out:
-                        f_out.write('\n'.join(self._resolve_content))
+                    # Do not overwrite if the hashes are the same but write if the file is empty
+                    # TODO: handle length 0 and empty contents
+                    if (undo_content == "" or hash_contents != unique_id) or not alias_file_exists:
```

Comment:
when initializing `hash_contents` with an empty string, I think we only need to compare `hash_contents` with `unique_id` here (as an empty string will also return an md5 value). Which in my opinion would lead to a structure like:

```
hash_contents = ''
alias_file_exists = os.path.isfile(self._filename_alias_content)
if alias_file_exists:
    try {...} except {}  # reading undo_content followed by hash_contents
```

In which case `hash_contents` is always empty if content or hash are either not valid or not available.